### PR TITLE
front: editor - change behavour of route tool 

### LINKED ISSUE #5903: front: refacto edition and creation forms for routes
# Description
For now, the route creation forms looks like this, with buttons allowing the user to select begin/end on the map:
image
When the begin/end have been selected, the tool proposes all the different possible routes. You can select one of them and specify if the transit is rigid or not. This is the process to create a new route. 

When you select a detector and you go in edition mode, you can see the routes linked to this detector. If you select one of these, you arrive to a form like this, with buttons to go to the begin/end detector, a delete button, a button to compute an alternative route and a button to change the transit from not rigid to rigid: 
image

**The only actions you can actually do on an existing route are deleting it or enable rigid transit (be careful you can not do the opposite).** 
Moreover, the creation and edition forms are very similar, so we'd like to merge them into one single form.

# Goal

1. Enable route edition
2. Avoid duplication in code and weird UI differences between both forms

# AC
- [x] Route edition and creation are merged into a single form displaing `Route` as a title.
- [x] When we click on the route tool the form contains the two selectors for the begin/end detectors
- [x] The following buttons for generic actions are showed as in the other object tools
image

- [x] When a detector is selected, there is a button to change it if we want (we dont need to delete it before)
- [x] Once the two detectors are entered, the button `Search for compatible routes` should be activated
- [x] When a route is selected, the button to choose if the transit is rigid or not will be activated
- [x] When we click on an existing route from the list in a detector form, the same route form is diplayed but in this case the button to delete is activated
- [x] `Search for alternative routes` in the current route edition tool will be renamed in #6124

### LINKED ISSUE #6142: front: buffer stops are sometimes called knocker
### What happened?

This is the route creation panel:

290868013-c6100b24-9c44-4993-808d-97a1471fc5ee

Buffer stops are called knockers.

### What did you expect to happen?

Bufferstops are called buffer stops.

### How can we reproduce it (as minimally and precisely as possible)?

Go to the route creation tool with the language set to english.

### What operating system, browser and environment are you using?

- Any

### OSRD version (top right corner `Account` button > `Informations`)

e3b89a8

### PR DESCRIPTION
**Pull request by sim51**
closes #5903
closes #6142

### CODE REVIEW

File: front/src/applications/editor/tools/trackEdition/tool.tsx [hunk 1]
```diff
@@ -1,21 +1,19 @@
...
+import { featureCollection } from '@turf/helpers';
 import nearestPointOnLine, { NearestPointOnLine } from '@turf/nearest-point-on-line';
 
 import { entityDoUpdate, getLineStringDistance } from 'common/IntervalsDataViz/data';
 import { ConfirmModal } from 'common/BootstrapSNCF/ModalSNCF';
 import { save } from 'reducers/editor';
 import { getMapMouseEventNearestFeature } from 'utils/mapHelper';
-
-import { NEW_ENTITY_ID } from '../../data/utils';
+import { NEW_ENTITY_ID } from 'applications/editor/data/utils';
```
Comment (SharglutDev) [inline front/src/applications/editor/tools/trackEdition/tool.tsx:17]: We decided to keep the relative imports when everything is inside the same module

File: front/src/applications/editor/tools/trackEdition/tool.tsx [hunk 2]
```diff
@@ -55,7 +53,7 @@ const TrackEditionTool: Tool<TrackEditionState> = {
     [
       {
         id: 'save-line',
-        icon: AiFillSave,
+        icon: MdShowChart,
```
Comment (SharglutDev) [inline front/src/applications/editor/tools/trackEdition/tool.tsx]: Why did you change this icon ?

File: front/src/applications/editor/tools/routeEdition/components/RouteEditionPanel.tsx [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+          })),
+        },
+      });
+    } catch (e) {
+      setState({ optionsState: { type: 'idle' } });
+    }
+  }, [infraID, setState, state.entity]);
+
+  /**
+   * Funciton that select a new route search item
```
Comment (SharglutDev) [inline front/src/applications/editor/tools/routeEdition/components/RouteEditionPanel.tsx]: Maybe clearer like that ?

```suggestion
   * Function that selects a new route search item or unselected it
```

File: front/src/applications/editor/tools/routeEdition/components/RouteEditionPanel.tsx [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+            },
+          },
+        };
+      });
+    },
+    [setState]
+  );
+
+  /**
+   * Function to sezarch for routes
```
Comment (SharglutDev) [inline front/src/applications/editor/tools/routeEdition/components/RouteEditionPanel.tsx]:
```suggestion
   * Function to search for routes
```

File: front/src/applications/editor/tools/routeEdition/components/RouteEditionPanel.tsx [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+
+  /**
+   * Funciton that select a new route search item
+   */
+  const selectSearchRouteItem = useCallback(
+    (index: number) => {
+      setState((prev) => {
+        if (prev.optionsState.type !== 'options' || !prev.entity) return prev;
+
+        const isSelection = index !== prev.optionsState.focusedOptionIndex;
```
Comment (SharglutDev) [inline front/src/applications/editor/tools/routeEdition/components/RouteEditionPanel.tsx]: I find it a bit weird that isSelection is true if the item is not selected. Here are some suggestions

```suggestion
        const isSelectionAction = index !== prev.optionsState.focusedOptionIndex;
        const isItemUnselected = index !== prev.optionsState.focusedOptionIndex;
        const isItemSelected = index === prev.optionsState.focusedOptionIndex;
```

**Review by SharglutDev — CHANGES_REQUESTED** [review]
Just started the review, will finish it tomorrow

File: front/src/applications/editor/tools/routeEdition/components/RouteMetadata.tsx [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        </span>
+        <div className="col-sm-6">
+          {!disabled && (
+            <span>
+              {t(
+                `Editor.tools.routes-edition.directions.${entity.properties.entry_point_direction}`
+              )}
+            </span>
+          )}
+        </div>
```
Comment (SharglutDev) [inline front/src/applications/editor/tools/routeEdition/components/RouteMetadata.tsx]: Maybe we can avoid rendering the div if `disabled`

```suggestion
     {!disabled && (
        <div className="col-sm-6">          
            <span>
              {t(
                `Editor.tools.routes-edition.directions.${entity.properties.entry_point_direction}`
              )}
            </span>
        </div>
        )}
```

File: front/src/applications/editor/tools/routeEdition/components/RouteEditionPanel.tsx [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+    (withDetector: boolean) => {
+      setState((prev) => {
+        let detectors: string[] = [];
+        if (
+          withDetector &&
+          prev.optionsState.type === 'options' &&
+          !isNil(prev.optionsState.focusedOptionIndex)
+        ) {
+          const selected = prev.optionsState.options[prev.optionsState.focusedOptionIndex];
+          detectors = selected.data.detectors;
```
Comment (SharglutDev) [inline front/src/applications/editor/tools/routeEdition/components/RouteEditionPanel.tsx]: Doesn't change much but might help for maintainability and harmonization with the other functions

```suggestion
          const selectedCandidate = prev.optionsState.options[prev.optionsState.focusedOptionIndex];
          detectors = selectedCandidate.data.detectors;
```

File: front/public/locales/en/translation.json
Comment (SharglutDev) [inline front/public/locales/en/translation.json:1]: For the 'no-waypoint-picked-yet' (l. 332) key, it should say bufferstop instead of knocker

File: front/src/applications/editor/tools/routeEdition/components/WayPointInput.tsx [hunk 8]
```diff
@@ -1,104 +1,96 @@
...
 import EditorContext from 'applications/editor/context';
+import { RouteEditionState, EndPointKeys } from 'applications/editor/tools/routeEdition/types';
 import { getEntity } from 'applications/editor/data/api';
-import Tipped from 'applications/editor/components/Tipped';
 import EntitySumUp from 'applications/editor/components/EntitySumUp';
-import { EndPointKeys } from 'applications/editor/tools/routeEdition/types';
-import type { EditRoutePathState } from 'applications/editor/tools/routeEdition/types';
-import type { ExtendedEditorContextType } from 'applications/editor/tools/editorContextTypes';
-
+import { EndPoint, WayPoint, WayPointEntity } from 'types';
```
Comment (SharglutDev) [inline front/src/applications/editor/tools/routeEdition/components/WayPointInput.tsx]:
```suggestion
import type { EndPoint, WayPoint, WayPointEntity } from 'types';
```

File: front/src/applications/editor/tools/routeEdition/components/WayPointInput.tsx [hunk 9]
```diff
@@ -1,104 +1,96 @@
...
-import { GoTrash } from 'react-icons/go';
+import { GoPencil } from 'react-icons/go';
 import { FaMapMarkedAlt, FaTimesCircle } from 'react-icons/fa';
-import type { Position } from 'geojson';
-
-import type { EndPoint, WayPoint, WayPointEntity } from 'types';
+import { isNil } from 'lodash';
 
 import EditorContext from 'applications/editor/context';
+import { RouteEditionState, EndPointKeys } from 'applications/editor/tools/routeEdition/types';
```
Comment (SharglutDev) [inline front/src/applications/editor/tools/routeEdition/components/WayPointInput.tsx]: See https://osrd.fr/en/docs/guides/contribute/conventions/front/#import--export

```suggestion
import { type RouteEditionState, EndPointKeys } from 'applications/editor/tools/routeEdition/types';
```

File: front/public/locales/fr/translation.json [hunk 10]
```diff
@@ -288,14 +288,16 @@
...
-          "pick-entrypoint-delete": "Supprimer le point de départ sélectionné",
-          "pick-exitpoint": "Outil de sélection du point d'arrivée",
-          "pick-exitpoint-cancel": "Annuler la sélection du point d'arrivée",
-          "pick-exitpoint-delete": "Supprimer le point d'arrivée sélectionné"
+          "pick-entry_point": "Outil de sélection du point de départ",
+          "pick-entry_point-cancel-delete": "Annuler la sélection du point de départ",
+          "pick-entry_point-delete": "Modify the selected starting point",
+          "pick-exit_point": "Outil de sélection du point d'arrivée",
+          "pick-exit_point-cancel-delete": "Annuler la sélection du point d'arrivée",
+          "pick-exit_point-delete": "Modify the selected starting point"
```
Comment (SharglutDev) [inline front/public/locales/fr/translation.json]:
```suggestion
          "pick-entry_point-delete": "Modifier le point de départ sélectionné",
          "pick-exit_point": "Outil de sélection du point d'arrivée",
          "pick-exit_point-cancel-delete": "Annuler la sélection du point d'arrivée",
          "pick-exit_point-delete": "Modifier le point de départ sélectionné"
```

File: front/src/applications/editor/tools/routeEdition/components/RouteEditionPanel.tsx [hunk 11]
```diff
@@ -0,0 +1,314 @@
...
+        entity: {
+          ...prev.entity,
+          properties: {
+            ...prev.entity.properties,
+            [EndPointKeys[endPoint]]: waypoint
+              ? {
+                  type: waypoint.objType,
+                  id: waypoint.properties.id,
+                }
+              : { type: 'Dector', id: NEW_ENTITY_ID },
```
Comment (SharglutDev) [inline front/src/applications/editor/tools/routeEdition/components/RouteEditionPanel.tsx]: I think ? There is no type safety ?

```suggestion
              : { type: 'Detector', id: NEW_ENTITY_ID },
```

File: front/src/applications/editor/tools/routeEdition/components/RouteMetadata.tsx [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+      </div>
+      <div className="row">
+        <label className="col-sm-6 col-form-label" htmlFor="include-release-detectors">
+          {t('Editor.tools.routes-edition.include-release-detectors')}
+        </label>
+        <div className="col-sm-6">
+          <input
+            type="checkbox"
+            id="include-release-detectors"
+            checked={disabled ? undefined : !!entity.properties.release_detectors.length}
```
Comment (SharglutDev) [inline front/src/applications/editor/tools/routeEdition/components/RouteMetadata.tsx]: When selecting a candidate for the first time, we have this warning in the console

Capture d’écran 2024-01-18 à 17 20 17

Is it necessary that the checkbox is at undefined if disabled ?

```suggestion
            checked={!!entity.properties.release_detectors.length}
```

File: front/src/applications/editor/tools/routeEdition/components/SearchRoute.tsx [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+  );
+  return (
+    <div>
+      <button
+        className="btn btn-primary btn-sm my-1 w-100"
+        disabled={state.type === 'loading' || disabled}
+        type="button"
+        onClick={searchFn}
+      >
+        <FiSearch /> {t(`Editor.tools.routes-edition.search-routes${!isNew ? '-alt' : ''}`)}
```
Comment (SharglutDev) [inline front/src/applications/editor/tools/routeEdition/components/SearchRoute.tsx]: Not in the docs but the new design system will include most of the github octicons so we try to use them as much as possible when they fit our needs

```suggestion
        <GoSearch /> {t(`Editor.tools.routes-edition.search-routes${!isNew ? '-alt' : ''}`)}
```

File: front/src/applications/editor/tools/routeEdition/components/RouteEditionPanel.tsx [hunk 14]
```diff
@@ -0,0 +1,314 @@
+import React, { FC, useCallback, useContext, useEffect, useMemo } from 'react';
+import { useDispatch } from 'react-redux';
+import { useTranslation } from 'react-i18next';
+import chroma from 'chroma-js';
+import { isNil } from 'lodash';
+
+import { EndPoint, RouteEntity, WayPointEntity } from 'types';
```
Comment (SharglutDev) [inline front/src/applications/editor/tools/routeEdition/components/RouteEditionPanel.tsx]:
```suggestion
import type { EndPoint, RouteEntity, WayPointEntity } from 'types';
```

File: front/src/applications/editor/tools/routeEdition/components/RouteEditionPanel.tsx [hunk 15]
```diff
@@ -0,0 +1,314 @@
...
+import EditorContext from 'applications/editor/context';
+import EntityError from 'applications/editor/components/EntityError';
+import { NEW_ENTITY_ID } from 'applications/editor/data/utils';
+import { osrdEditoastApi } from 'common/api/osrdEditoastApi';
+import { useInfraID } from 'common/osrdContext';
+import { save } from 'reducers/editor';
+import { EndPointKeys, RouteEditionState } from '../types';
```
Comment (SharglutDev) [inline front/src/applications/editor/tools/routeEdition/components/RouteEditionPanel.tsx]:
```suggestion
import { EndPointKeys, type RouteEditionState } from '../types';
```

File: front/src/applications/editor/tools/routeEdition/components/RouteEditionPanel.tsx [hunk 16]
```diff
@@ -0,0 +1,314 @@
...
+import { ExtendedEditorContextType } from '../../editorContextTypes';
```
Comment (SharglutDev) [inline front/src/applications/editor/tools/routeEdition/components/RouteEditionPanel.tsx]:
```suggestion
import type { ExtendedEditorContextType } from '../../editorContextTypes';
```

File: front/src/applications/editor/tools/routeEdition/components/SearchRoute.tsx [hunk 17]
```diff
@@ -0,0 +1,66 @@
+import React, { FC, useMemo } from 'react';
+import { useTranslation } from 'react-i18next';
+import { FiSearch } from 'react-icons/fi';
+
+import { LoaderFill } from 'common/Loaders';
+import { SearchRouteItem } from './SearchRouteItem';
+import { OptionsStateType } from '../types';
```
Comment (SharglutDev) [inline front/src/applications/editor/tools/routeEdition/components/SearchRoute.tsx]:
```suggestion
import type { OptionsStateType } from '../types';
```

File: front/src/applications/editor/tools/routeEdition/components/RouteMetadata.tsx [hunk 18]
```diff
@@ -0,0 +1,58 @@
+import React, { FC } from 'react';
+import { useTranslation } from 'react-i18next';
+
+import { RouteEntity } from 'types';
```
Comment (SharglutDev) [inline front/src/applications/editor/tools/routeEdition/components/RouteMetadata.tsx]:
```suggestion
import type { RouteEntity } from 'types';
```

File: front/src/applications/editor/tools/routeEdition/components/EndPoints.tsx [hunk 19]
```diff
@@ -0,0 +1,64 @@
+import React, { FC, useMemo } from 'react';
+import { BsArrowBarRight } from 'react-icons/bs';
+import { HiSwitchVertical } from 'react-icons/hi';
+import { FaFlagCheckered } from 'react-icons/fa';
+import { useTranslation } from 'react-i18next';
+
+import { EndPoint, WayPointEntity } from 'types';
+import { RouteEditionState } from 'applications/editor/tools/routeEdition/types';
```
Comment (SharglutDev) [inline front/src/applications/editor/tools/routeEdition/components/EndPoints.tsx]:
```suggestion
import type { EndPoint, WayPointEntity } from 'types';
import type { RouteEditionState } from 'applications/editor/tools/routeEdition/types';
```

**Review by SharglutDev — CHANGES_REQUESTED** [review]
Second wave of comments, still not finished sorry, had a lot to do today

File: front/src/applications/editor/tools/routeEdition/components/SearchRouteItem.tsx [hunk 20]
```diff
@@ -0,0 +1,58 @@
+import React, { FC } from 'react';
+import { useTranslation } from 'react-i18next';
+import cx from 'classnames';
+
+import { RouteCandidate } from '../types';
```
Comment (SharglutDev) [inline front/src/applications/editor/tools/routeEdition/components/SearchRouteItem.tsx]:
```suggestion
import type { RouteCandidate } from '../types';
```

File: front/src/applications/editor/tools/routeEdition/components.tsx [hunk 21]
```diff
@@ -1,66 +1,51 @@
...
-import {
-  EditRoutePathEditionLayers,
-  EditRoutePathLeftPanel,
-} from 'applications/editor/tools/routeEdition/components/EditRoutePath';
-import {
-  EditRouteMetadataLayers,
-  EditRouteMetadataPanel,
-} from 'applications/editor/tools/routeEdition/components/EditRouteMetadata';
-
+import { ExtendedEditorContextType } from 'applications/editor/tools/editorContextTypes';
```
Comment (SharglutDev) [inline front/src/applications/editor/tools/routeEdition/components.tsx]:
```suggestion
import type { ExtendedEditorContextType } from 'applications/editor/tools/editorContextTypes';
```

File: front/src/applications/editor/tools/routeEdition/components.tsx [hunk 22]
```diff
@@ -1,66 +1,51 @@
...
+import { NEW_ENTITY_ID } from 'applications/editor/data/utils';
 import colors from 'common/Map/Consts/colors';
 import GeoJSONs from 'common/Map/Layers/GeoJSONs';
-import { useInfraID } from 'common/osrdContext';
-
 import { getMap } from 'reducers/map/selectors';
+import { useInfraID } from 'common/osrdContext';
+import { RouteEditionState } from './types';
```
Comment (SharglutDev) [inline front/src/applications/editor/tools/routeEdition/components.tsx]:
```suggestion
import type { RouteEditionState } from './types';
```

File: front/src/applications/editor/tools/routeEdition/tool.tsx [hunk 23]
```diff
@@ -0,0 +1,120 @@
...
+      },
+    ],
+    [
+      {
+        id: 'delete-route',
+        icon: GoTrash,
+        labelTranslationKey: 'Editor.tools.routes-edition.actions.delete-route',
+        // Show button only if we are editing
+        isDisabled({ state: { entity } }) {
+          return !entity?.properties.id;
```
Comment (SharglutDev) [inline front/src/applications/editor/tools/routeEdition/tool.tsx]: The only case I see when we would want to disable the delete would be when creating a new route from scratch. What do you think ?

```suggestion
        isDisabled({ state: { initialEntity } }) {
          return !initialEntity;
```

File: front/src/applications/editor/tools/routeEdition/tool.tsx [hunk 24]
```diff
@@ -0,0 +1,120 @@
...
+                  await dispatch<ReturnType<typeof save>>(
+                    save(infraID, { delete: [state.entity] })
+                  );
+                  setState(getRouteEditionState());
+                  closeModal();
+                  forceRender();
+                }
+              }}
+            >
+              <p>{t('Editor.tools.routes-edition.delete-route-success').toString()}</p>
```
Comment (SharglutDev) [inline front/src/applications/editor/tools/routeEdition/tool.tsx]:
```suggestion
              <p>{t('Editor.tools.routes-edition.confirm-delete-route').toString()}</p>
```

File: front/src/applications/editor/tools/routeEdition/tool.tsx [hunk 25]
```diff
@@ -0,0 +1,120 @@
...
+    if (state.extremityState.type === 'selection' && state.hovered) return 'pointer';
+    return 'default';
+  },
+  onKeyDown(e, { state, setState }) {
+    if (state.extremityState.type === 'selection' && e.key === 'Escape') {
+      setState({
+        ...state,
+        extremityState: { type: 'idle' },
+      });
+    }
```
Comment (SharglutDev) [inline front/src/applications/editor/tools/routeEdition/tool.tsx]: The short key doesn't seem to work

**Review by SharglutDev — CHANGES_REQUESTED** [review]
Thank you for your PR. Tested and it works great, most of the comments are for syntax and code readability.

File: front/src/applications/editor/tools/trackEdition/tool.tsx (hunk 2 above)
Comment (sim51) [reply front/src/applications/editor/tools/trackEdition/tool.tsx]: Don't know ... I think I did a mistake here

**Review by clarani — CHANGES_REQUESTED** [review]
I found a bug: if I create a route, and then I change the departure or exit point and save, the route is not correctly displayed with the dotted line 
image

**Review by SharglutDev — CHANGES_REQUESTED** [review]
Some extra comments after testing the app with @bgiuliana : 

- For the info text of "rigid transit", it should say : "Cette propriété ne peut pas être réactivée après la sauvegarde, et peut uniquement être désactivée." instead of "a posteriori" for french translation and "after saving" in english instead of "afterwards".

- the current behavior for transit rigit checked is wrong (see https://github.com/osrd-project/osrd/issues/6188). It should still be fixed in a different PR but ideally it would be great if it was this week too (as the bug has been upgraded to major)

**Comment by sim51** [comment]
I changed the code, based on your review.

I still have to do : 
- the `escape` shortcut
- fix the clara issue about the display of route when you switch

File: front/src/applications/editor/tools/routeEdition/components/RouteMetadata.tsx (hunk 12 above)
Comment (sim51) [reply front/src/applications/editor/tools/routeEdition/components/RouteMetadata.tsx]: It's just to indicate to the user that there is no data. 
So when a user select a route, there is a visual diff (in addition of the opacity ).

File: front/src/applications/editor/tools/routeEdition/tool.tsx (hunk 25 above)
Comment (sim51) [reply front/src/applications/editor/tools/routeEdition/tool.tsx]: For this one, it's a good cacth !
The `onKeyPress` is registered on the map, not on the window ... 
So if you click somewhere on the map and then press`escape` it works.

**Comment by sim51** [comment]
About @clarani 's issue , when user changes a "way point" we can't display a line on the map.
We have geometries only for the initial entity for "searched route".

So the code issue was due a geometry set to null, which don't trigger a re-render of the map component.  I replace it by an empty geometry.

**Review by SharglutDev — APPROVED** [review]
Lgtm and tested (@clarani bug is fixed as well). Great work ! :)

**Review by clarani — APPROVED** [review]
Bug is fixed :white_check_mark: I didn't check the code
